### PR TITLE
Update index_data.py for datatype conversion and alignment

### PR DESCRIPTION
**Pull request by taozhiwang**
## Description

Hi,

Here I change `index_data.py` for converting time in datetime64 into the same precision with user data. This helps users avoid query failures due to precision issues.

Datetype detection is performed during initialization to avoid inconsistencies in the format of user data. This may have some impact on efficiency.

**Changed files**:
1. `index_data.py` change `__init__` under `Index` for the initial alignment test, also change `_convert_type` for detecting data type and convert precision
3. `test_index_data.py` test above changes
4. `shrink.py` pre-commit auto-correction

## Motivation and Context

[issue 1806](https://github.com/microsoft/qlib/issues/1806)

## How Has This Been Tested?

- [x] Pass the test by running: `pytest qlib/tests/test_all_pipeline.py` under upper directory of `qlib`.
- [ ] If you are adding a new feature, test on your own test scripts.

## Screenshots of Test Results (if appropriate):
1. Pipeline test:
image

3. Your own tests:
image

## Types of changes

- [x] Fix bugs
- [ ] Add new feature
- [ ] Update documentation

### CODE REVIEW

**Comment by taozhiwang** [comment]
@microsoft-github-policy-service agree

File: qlib/utils/index_data.py [hunk 1]
```diff
@@ -131,7 +137,11 @@ def _convert_type(self, item):
         if self.idx_list.dtype.type is np.datetime64:
             if isinstance(item, pd.Timestamp):
                 # This happens often when creating index based on pandas.DatetimeIndex and query with pd.Timestamp
-                return item.to_numpy()
+                return item.to_numpy().astype(self.idx_list.dtype)
+            elif isinstance(item, np.datetime64):
+                # This happens often when creating index based on np.datetime64 and query with another precision
+                return item.astype(self.idx_list.dtype)
+
```
Comment (you-n-g) [inline qlib/utils/index_data.py:144]:
```suggestion
            # NOTE: It is hard to consider every cases at first. 
            # We just try to cover part of cases to make it more user-friendly 
```

File: qlib/utils/index_data.py [hunk 2]
```diff
@@ -131,7 +137,12 @@ def _convert_type(self, item):
         if self.idx_list.dtype.type is np.datetime64:
             if isinstance(item, pd.Timestamp):
                 # This happens often when creating index based on pandas.DatetimeIndex and query with pd.Timestamp
-                return item.to_numpy()
+                return item.to_numpy().astype(self.idx_list.dtype)
+            elif isinstance(item, np.datetime64):
+                # This happens often when creating index based on np.datetime64 and query with another precision
+                return item.astype(self.idx_list.dtype)
+            # NOTE: It is hard to consider every cases at first.
```
Comment (you-n-g) [inline qlib/utils/index_data.py:144]:
```suggestion
            # NOTE: It is hard to consider every case at first. 
```

**Comment by you-n-g** [comment]
Thanks for your great efforts.

**Pull request merged**